### PR TITLE
Add removeRange API to DbProvider

### PR DESCRIPTION
**Pull request by michtsu**
This change adds a removeRange API to DbProvider. This allows removal of items by specifying a low/high range, similar to the arguments of the getRange() API.

### CODE REVIEW

File: src/InMemoryProvider.ts [hunk 1]
```diff
@@ -256,6 +271,19 @@ class InMemoryStore implements DbStore {
         this._mergedData = {};
         return Promise.resolve<void>(void 0);
     }
+
+    public removeEx(keys: string[]): Promise<void> {
```
Comment (masterkidan) [inline src/InMemoryProvider.ts]: Make this private?

File: src/InMemoryProvider.ts [hunk 2]
```diff
@@ -402,4 +430,29 @@ class InMemoryIndex extends DbIndexFTSFromRangeQueries {
...
         return Promise.resolve(keys.length);
     }
+
+    removeRange(keyLowRange: KeyType, 
+                keyHighRange: KeyType, 
+                lowRangeExclusive?: boolean, 
+                highRangeExclusive?: boolean): Promise<void> {
+        // not implemented - see InMemoryProvider.removeRange
+        return Promise.reject(void 0);
+    }
```
Comment (masterkidan) [inline src/InMemoryProvider.ts]: Probably no longer required?

File: src/NoSqlProvider.ts [hunk 3]
```diff
@@ -181,6 +182,17 @@ export abstract class DbProvider {
...
     }
 
+    removeRange(storeName: string,
+                indexName: string,
+                keyLowRange: KeyType,
+                keyHighRange: KeyType,
+                lowRangeExclusive?: boolean,
+                highRangeExclusive?: boolean): Promise<void> {
+      return this._getStoreIndexTransaction(storeName, true, indexName).then( index => {
+        return index.removeRange(keyLowRange, keyHighRange, lowRangeExclusive, highRangeExclusive);
```
Comment (masterkidan) [inline src/NoSqlProvider.ts]: can this also not be store .removeRange?

File: src/InMemoryProvider.ts (hunk 2 above)
Comment (michtsu) [reply src/InMemoryProvider.ts]: The API is defined on DbIndex interface - it *is* implemented in IndexedDbIndex.  It's just that for InMemoryIndex, it didn't make sense to implement.

File: src/InMemoryProvider.ts [hunk 4]
```diff
@@ -402,4 +428,38 @@ class InMemoryIndex extends DbIndexFTSFromRangeQueries {
...
+        if (isError(keys)) {
+            return Promise.reject(void 0);
+        }
+        const store = attempt(() => {
+            return this._trans.getStore(storeName);
+        });
+        if (isError(store)) {
+            return Promise.reject(void 0);
+        }
+        return (store as InMemoryStore).removeEx(keys);
```
[the remark this answers is not in the document]
Comment (michtsu) [reply src/InMemoryProvider.ts]: Please see my latest commit - I think you were looking at an older iteration :)

File: src/InMemoryProvider.ts [hunk 5]
```diff
@@ -54,7 +54,19 @@ export class InMemoryProvider extends DbProvider {
...
-    }
+    }     
+
+    removeRange(storeName: string,
+                indexName: string,
+                keyLowRange: KeyType,
+                keyHighRange: KeyType,
+                lowRangeExclusive?: boolean,
+                highRangeExclusive?: boolean): Promise<void> {
+        return this._getStoreIndexTransaction(storeName, true, indexName).then( index => {
```
[the remark this answers is not in the document]
Comment (michtsu) [reply src/InMemoryProvider.ts]: Yes, please see my latest iteration

File: src/InMemoryProvider.ts (hunk 1 above)
Comment (michtsu) [reply src/InMemoryProvider.ts]: Yes

File: src/InMemoryProvider.ts (hunk 2 above)
Comment (michtsu) [reply src/InMemoryProvider.ts]: Actually, I will move removeRange to the store API as you suggested - then I can remove this.

File: src/NoSqlProvider.ts (hunk 3 above)
Comment (michtsu) [reply src/NoSqlProvider.ts]: Yes - will make this change

File: src/tests/NoSqlProviderTests.ts [hunk 6]
```diff
@@ -485,9 +485,71 @@ describe('NoSqlProvider', function () {
...
+                                        assert.equal(rets[0].id, 'a1');
+                                        assert.equal(rets[1].id, 'a10');
+                                        assert.equal(rets[2].id, 'a2');
+                                        assert.equal(rets[3].id, 'a3');
+                                        assert.equal(rets[4].id, 'a7');
+                                        assert.equal(rets[5].id, 'a8');
+                                        assert.equal(rets[6].id, 'a9');
+                                        return Promise.resolve().then(() => {
+                                            // Test removing low exclusive, high inclusive
+                                            return prov.removeRange('test', '', 'a7', 'a9', true, false).then(() => {
```
Comment (masterkidan) [inline src/tests/NoSqlProviderTests.ts]: Can you make this a separate test?

File: src/tests/NoSqlProviderTests.ts [hunk 7]
```diff
@@ -485,9 +485,71 @@ describe('NoSqlProvider', function () {
...
+                                                    assert(!!rets);
+                                                    assert.equal(rets.length, 5);
+                                                    assert.equal(rets[0].id, 'a1');
+                                                    assert.equal(rets[1].id, 'a10');        
+                                                    assert.equal(rets[2].id, 'a2');
+                                                    assert.equal(rets[3].id, 'a3');
+                                                    assert.equal(rets[4].id, 'a7');
+                                                    return Promise.resolve().then(() => {
+                                                        // Test removing low inclusive, high exclusive
+                                                        return prov.removeRange('test', '', 'a1', 'a3', false, true).then(() => {
```
Comment (masterkidan) [inline src/tests/NoSqlProviderTests.ts]: Same here , make this a separate test as well

File: src/InMemoryProvider.ts [hunk 8]
```diff
@@ -54,7 +54,7 @@ export class InMemoryProvider extends DbProvider {
 
     internal_getStore(name: string): StoreData {
         return this._stores[name];
-    }
+    }
```
Comment (masterkidan) [inline src/InMemoryProvider.ts]: nit: unwanted whitespace

File: src/InMemoryProvider.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (michtsu) [reply src/InMemoryProvider.ts]: Yes

File: src/InMemoryProvider.ts [hunk 9]
```diff
@@ -256,6 +272,19 @@ class InMemoryStore implements DbStore {
         this._mergedData = {};
         return Promise.resolve<void>(void 0);
     }
+
+    private removeEx(keys: string[]): Promise<void> {
```
Comment (masterkidan) [inline src/InMemoryProvider.ts]: nit: Rename to removeInternal, or _remove

File: src/IndexedDbProvider.ts [hunk 10]
```diff
@@ -815,6 +829,22 @@ class IndexedDbIndex extends DbIndexFTSFromRangeQueries {
         return this._countRequest(req);
     }
 
+    public removeRange(keyLowRange: KeyType,
```
Comment (masterkidan) [inline src/IndexedDbProvider.ts]: Can you change this to instead be getKeysForIndex?

File: src/IndexedDbProvider.ts [hunk 11]
```diff
@@ -659,6 +659,20 @@ class IndexedDbStore implements DbStore {
...
+                keyHighRange: KeyType, 
+                lowRangeExclusive?: boolean, 
+                highRangeExclusive?: boolean): Promise<void> {
+        const index = attempt(() => {
+            return indexName ? this.openIndex(indexName) : this.openPrimaryKey();
+        });
+        if (!index || isError(index)) {
+            return Promise.reject<void>('Index "' + indexName + '" not found');
+        }                    
+        return (index as IndexedDbIndex).removeRange(keyLowRange, keyHighRange, lowRangeExclusive, highRangeExclusive);
```
Comment (masterkidan) [inline src/IndexedDbProvider.ts]: Can you follow the same pattern used in the inMemory provider? 
Basically use the index to just generate the keys and then remove them from the store instead?

The current remove method doesn't take into account the fake indexes and fulttext indexes that are created.. might be easier to get the keys from the index and reuse the existing remove method.

File: src/InMemoryProvider.ts (hunk 8 above)
Comment (michtsu) [reply src/InMemoryProvider.ts]: Removed

File: src/InMemoryProvider.ts (hunk 9 above)
Comment (michtsu) [reply src/InMemoryProvider.ts]: Changed to removeInternal

File: src/IndexedDbProvider.ts (hunk 10 above)
Comment (michtsu) [reply src/IndexedDbProvider.ts]: Changed to getKeysForRange

File: src/IndexedDbProvider.ts (hunk 11 above)
Comment (michtsu) [reply src/IndexedDbProvider.ts]: Made this change

File: src/InMemoryProvider.ts [hunk 12]
```diff
@@ -342,6 +368,21 @@ class InMemoryIndex extends DbIndexFTSFromRangeQueries {
...
+        : Promise<any[]> {
+        if (!this._trans.internal_isOpen()) {
+            return Promise.reject('InMemoryTransaction already closed');
+        }
+        const keys = attempt(() => {
+            const data = this._calcChunkedData();
+            return this._getKeysForRange(data, keyLowRange, keyHighRange, lowRangeExclusive, highRangeExclusive);
+        });
+        if (isError(keys)) {
+            return Promise.reject(void 0);
```
Comment (masterkidan) [inline src/InMemoryProvider.ts]: Probably bubble the error up, rather than setting void 0

File: src/IndexedDbProvider.ts [hunk 13]
```diff
@@ -815,6 +831,27 @@ class IndexedDbIndex extends DbIndexFTSFromRangeQueries {
...
+            return this._getKeyRangeForRange(keyLowRange, keyHighRange, lowRangeExclusive, highRangeExclusive);
+        });
+        if (isError(keyRange)) {
+            return Promise.reject<string[]>(keyRange);
+        }
+        if (this._store.getAllKeys && !this._fakeComplicatedKeys) {
+            return IndexedDbProvider.WrapRequest<any[]>(this._store.getAllKeys(keyRange));
+        }
+
+        let keys: any[] = [];
```
Comment (masterkidan) [inline src/IndexedDbProvider.ts]: When fakeComplicatedKeys is enabled, we may need to actually look up a different store, and not operate on the current store ... 
Can you account for that fact?

File: src/NoSqlProvider.ts [hunk 14]
```diff
@@ -144,7 +150,7 @@ export abstract class DbProvider {
 
     protected abstract _deleteDatabaseInternal(): Promise<void>;
 
-    private _getStoreTransaction(storeName: string, readWrite: boolean): Promise<DbStore> {
+    protected _getStoreTransaction(storeName: string, readWrite: boolean): Promise<DbStore> {
```
Comment (masterkidan) [inline src/NoSqlProvider.ts]: Can this be changed back to private.

File: src/tests/NoSqlProviderTests.ts (hunk 6 above)
Comment (michtsu) [reply src/tests/NoSqlProviderTests.ts]: Split out, and added more tests

File: src/tests/NoSqlProviderTests.ts (hunk 7 above)
Comment (michtsu) [reply src/tests/NoSqlProviderTests.ts]: Split out, and added more tests

File: src/InMemoryProvider.ts (hunk 12 above)
Comment (masterkidan) [reply src/InMemoryProvider.ts]: Actually cancel this, I see that the rest of the code also does the same thing.

File: src/IndexedDbProvider.ts (hunk 13 above)
Comment (masterkidan) [reply src/IndexedDbProvider.ts]: This is because the index itself may be in a separate store.

Comment (michtsu) [reply src/IndexedDbProvider.ts]: As discussed, looks like openIndex handles this.

File: src/NoSqlProvider.ts (hunk 14 above)
Comment (michtsu) [reply src/NoSqlProvider.ts]: yes

**Pull request merged**